.s2i/bin/assemble

```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
+
+ASSEMBLE_SCRIPT="/usr/libexec/s2i/assemble"
+
+pushd /tmp/src || exit 1
+
+echo "---> Building your Ruby application from source ..."
+if [ -f Gemfile ]; then
+  ADDTL_BUNDLE_ARGS="--retry 2"
+  if [ -f Gemfile.lock ]; then
+    # This is new in version 3.0
+    # [DEPRECATED] The `--deployment` flag is deprecated because it relies
+    # on being remembered across bundler invocations,
+    # which bundler will no longer do in future versions.
+    # Instead please use `bundle config set --local deployment 'true'`,
+    # and stop using this flag
+    bundle config set --local deployment 'true'
+  fi
+
+  if [[ "$RAILS_ENV" == "development" || "$RACK_ENV" == "development" ]]; then
+    BUNDLE_WITHOUT=${BUNDLE_WITHOUT:-"test"}
+  elif [[ "$RAILS_ENV" == "test" || "$RACK_ENV" == "test" ]]; then
+    BUNDLE_WITHOUT=${BUNDLE_WITHOUT:-"development"}
+  else
+    BUNDLE_WITHOUT=${BUNDLE_WITHOUT:-"development:test"}
+  fi
+
+  if [ -n "$BUNDLE_WITHOUT" ]; then
+    # [DEPRECATED] The `--without` flag is deprecated because
+    # it relies on being remembered across bundler invocations,
+    # which bundler will no longer do in future versions.
+    # Instead please use `bundle config set --local without ''test''`,
+    # and stop using this flag
+    bundle config set --local without '$BUNDLE_WITHOUT'
+  fi
+
+  echo "---> Running 'bundle install ${ADDTL_BUNDLE_ARGS}' ..."
+  # [DEPRECATED] The `--path` flag is deprecated because
+  # it relies on being remembered across bundler invocations,
+  # which bundler will no longer do in future versions.
+  # Instead please use `bundle config set --local path './bundle'`,
+  # and stop using this flag
+  bundle config set --local path './bundle'
+  bundle install ${ADDTL_BUNDLE_ARGS}
+
+  echo "---> Cleaning up unused ruby gems ..."
+  bundle clean --verbose
+fi
+
+if [[ "$RAILS_MASTER_KEY" == "" && ! $(ls config/master.key 2>/dev/null 1>/dev/null) ]]; then
+	if [[ $(ls config/secrets.yml.enc 2>/dev/null 1>/dev/null) ]]; then
+		# If the master.key was regenerated while there is already
+		# an encoded file with secrets can only be decoded with the original key.
+		echo "ERROR: no key detected, but encoded file is present!"
+		exit 1
+	fi
+
+	echo "---> WARNING: No master key present in environment, generating..."
+	EDITOR=/bin/true bin/rails credentials:edit
+fi
+
+popd
+
+# Application cannot be moved without this cleanup.
```

Comment:
Moved? What do you mean?

Reply:
Not quite sure to be honest.

The original assemble would fail on this line https://github.com/sclorg/s2i-ruby-container/blob/master/3.0/s2i/bin/assemble#L24
complaining about .bundle not being empty in dest (because we already have .bundle in the /tmp/src from where we are copying and `~/.bundle` was created by bundler, unfortunately for us, the "home" directory is also the destination for s2i)
not sure I ever encountered this kind of error using `mv`.

Reply:
Ah, I see. So I think we're duplicating some logic here. There is to be only one `.bundle`, not two of them... wiping the bundle directory would mean it needs to be re-downloaded and reinstalled :/.  Any chance on having only one? Maybe this should be a ticket for `s2i-ruby-container` ... If the `mv` would have `-n`, `-u` or `-f` I suspect this would not happen.

Reply:
>  Any chance on having only one?

Hmm... There might be a bundler option for that but I wouldn't bet on it.

>wiping the bundle directory would mean it needs to be re-downloaded and reinstalled

The gems are already installed in the `/tmp/src/bundle` not `/tmp/src/.bundle` (that's where the options we set are, notice the dot) before the move. We do not touch that. What is generating the conflict is the options for bundler. We have 2 of them. One "global" (specific to the user), one local (specific to the folder).

So we leave the already downloaded and installed gems untouched (and running the s2i build does not cause re-download https://gist.github.com/jackorp/8bdec3c51e84d7f93c13dfb30b32c559 ).

AFAICT by removing the `~/.bundle`, we delete some caches and global options (such as default gem test suite, CI, linter, etc., used when creating a gem). Not important to us and then move the downloaded gems with the whole app, thanks to the entire bundle logic copied over, so we don't re-download everything.

To clarify, the first bundle download invocation creates 2 `.bundle` directories:
  * In the home directory (here, the `/opt/app-src/...` directory where the sources are moved to in the end), holding global cache and some global options, we don't care about that much.
  * In the local directory where it is run, containing local options, we set with `bundle config set --local <...>`, which is important to us
  
> If the mv would have -n, -u or -f I suspect this would not happen.

Good idea. I suspect we want to use `-f`, I'll open the ticket.

Reply:
Alternatively, we can remove the `.bundle` before running `popd` as a way to cleanup after us. We also retain the .bundle in the `/opt/app-root/src`, and options are being set again later anyway. No redownloads were triggered either way. https://gist.github.com/jackorp/722e4f5c9a8b95562b1f0b2597a02e8e

Reply:
I see. I thought it's the installation folder. Feel free to remove it then.